augly/text/augmenters/typo.py

```diff
@@ -149,23 +154,33 @@ def substitute(self, data: str) -> str:
             get_aug_idxes(self, tokens, filtered_word_idxes, aug_word_cnt, Method.WORD)
         )
 
-        for t_i, token in enumerate(tokens):
-            if t_i not in aug_word_idxes:
-                results.append(token)
-                continue
-
-            misspellings = self.model.replace(token) if self.model else None
-            if misspellings:
-                misspelling = self.sample(misspellings, 1)[0]
-                results.append(self.align_capitalization(token, misspelling))
-            elif len(self.augmenters) > 0:
-                aug = self.sample(self.augmenters, 1)[0]
-                new_token = aug.augment(token)
-                results.append(self.align_capitalization(token, new_token))
-            else:
-                # If no misspelling is found in the dict & no other typo types are being
-                # used, don't change the token
-                results.append(token)
+        for t_i in range(1, self.max_typo_length + 1):
```

Comment:
since `max_typo_length` is only really relevant if the `misspelling` dictionary is being used, i wonder if it makes more sense for the loops to be in reverse, i.e.

instead of iterating on the max typo length and THEN iterating on the tokens, i think it makes more sense to do something like:

```
# iterate over tokens, could be while or for loop
     if self.model is not None:  # ie, mispelling dict is being used
         # loop on mispelling length, and choose a mispelling to act on
     elif .....
```

does that make any sense / do you agree? i feel like since this is only affecting misspelling dict, it shouldn't affect the other types too. (cc: @zpapakipos)

Reply:
I see your point & would be fine with it being implemented that way, but I'm also fine if we keep it this way and think it might end up being cleaner.

Note that Erik is setting `self.max_typo_length` to 1 if the typo type doesn't include misspelling, so if there are no misspellings the outer loop here will only execute once (`for t_i in range(1, 2)`), so we don't actually waste any compute time. Also, if we only did the `for t_i in range(1, self.max_typo_length + 1)` loop if `self.model is not None`, then we would end up with some duplicate code since we would basically need this whole loop inside that if statement, but if `self.model is None` then we would need the same loop but without the `for t_i in range(1, self.max_typo_length + 1)` loop. So even though I see your point, I feel like it might be simpler & cleaner to leave it like this. Lmk your thoughts!